File: eclipse/src/hu/webarticum/gitchordgui/eclipse/services/GitChordService.java
package hu.webarticum.gitchordgui.eclipse.services;

import hu.webarticum.gitchordgui.eclipse.model.CommandResultPayload;
import hu.webarticum.gitchordgui.eclipse.model.PageIntent;
import hu.webarticum.gitchordgui.eclipse.model.PageOpenRequest;
import hu.webarticum.gitchordgui.eclipse.model.PanelGroup;
import hu.webarticum.gitchordgui.eclipse.model.PanelState;
import hu.webarticum.gitchordgui.eclipse.preferences.GitChordPreferences;
import hu.webarticum.gitchordgui.eclipse.util.JsonUtil;
import hu.webarticum.gitchordgui.eclipse.views.GitChordGui;
import java.io.File;
import java.io.InputStream;
import java.net.URI;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.regex.Pattern;
import org.eclipse.core.filesystem.EFS;
import org.eclipse.core.resources.IResource;
import org.eclipse.core.resources.ResourcesPlugin;
import org.eclipse.core.runtime.IAdaptable;
import org.eclipse.core.runtime.IPath;
import org.eclipse.jface.dialogs.MessageDialog;
import org.eclipse.jface.viewers.ISelection;
import org.eclipse.jface.viewers.IStructuredSelection;
import org.eclipse.swt.widgets.Display;
import org.eclipse.ui.IEditorInput;
import org.eclipse.ui.IEditorPart;
import org.eclipse.ui.IPathEditorInput;
import org.eclipse.ui.IURIEditorInput;
import org.eclipse.ui.IViewPart;
import org.eclipse.ui.IWorkbenchPage;
import org.eclipse.ui.IWorkbenchWindow;
import org.eclipse.ui.PartInitException;
import org.eclipse.ui.PlatformUI;
import org.eclipse.ui.ide.IDE;

public final class GitChordService {

	private static final String GLOBAL_SECONDARY_ID = "";
	private static final long COMMAND_TIMEOUT_MILLIS = 30_000L;
	private static final long COMMIT_TIMEOUT_MILLIS = 30_000L;
	private static final long REPO_RESOLVE_TIMEOUT_MILLIS = 5_000L;
	private static final Pattern COMMIT_ID_PATTERN = Pattern.compile("^[0-9a-fA-F]{7,40}$");

	private static final GitChordService INSTANCE = new GitChordService();

	private final ExecutorService executorService = Executors.newCachedThreadPool();

	private final Map<String, PanelState> statesBySecondaryId = new LinkedHashMap<>();

	private final Map<String, GitChordGui> viewsByGroupKey = new LinkedHashMap<>();

	private GitChordService() {
		// Singleton.
	}

	public static GitChordService getInstance() {
		return INSTANCE;
	}

	public synchronized PanelState attachView(GitChordGui view, String secondaryId) {
		String normalizedSecondaryId = normalizeSecondaryId(secondaryId);
		PanelState state = statesBySecondaryId.computeIfAbsent(
			normalizedSecondaryId,
			id -> new PanelState(PanelGroup.global(), "/about", null)
		);
		viewsByGroupKey.put(state.group().key(), view);
		return state;
	}

	public synchronized void detachView(GitChordGui view) {
		viewsByGroupKey.values().removeIf(candidate -> candidate == view);
	}

	public void openGlobalPage(String path) {
		openPage(PanelGroup.global(), path, null);
	}

	public void openCurrentRepoPage(String path) {
		openCurrentRepoPage(path, null);
	}

	public void openCurrentRepoPage(String path, PageIntent intent) {
		runInBackground(() -> {
			String repoRoot = resolveCurrentRepoRoot();
			runOnUiThread(() -> {
				if (repoRoot == null) {
					showError("No Git repository context found.");
					openPage(PanelGroup.global(), "/".equals(path) ? "/repo-state" : path, intent);
				} else {
					openPage(PanelGroup.repo(repoRoot), path, intent);
				}
			});
		});
	}

	public void openRepoStateForSelection(ISelection selection) {
		String candidatePath = candidatePathForSelection(selection);
		if (candidatePath == null) {
			showError("No file-system resource was selected.");
			return;
		}

		runInBackground(() -> {
			String repoRoot = resolveRepoRootForPath(candidatePath);
			runOnUiThread(() -> {
				if (repoRoot == null) {
					showError("No Git repository found for the selected resource.");
				} else {
					openPage(PanelGroup.repo(repoRoot), "/", null);
				}
			});
		});
	}

	public void handleViewMessage(GitChordGui view, String rawMessage) {
		Map<String, Object> message = JsonUtil.parseObject(rawMessage);
		String type = asString(message.get("type"));
		if (type == null) {
			return;
		}

		switch (type) {
			case "requestInitialState":
				view.sendInitialState(initialRepoRoot(view.group()), GitChordPreferences.effectiveLanguage());
				break;
			case "requestRepoContext":
				refreshViewRepoContext(view);
				break;
			case "openPage":
				handleOpenPageRequest(view, PageOpenRequest.fromPayload(message.get("request")));
				break;
			case "setLanguage":
				handleSetLanguage(asString(message.get("language")));
				break;
			case "openCommit":
				handleOpenCommit(view.group(), asString(message.get("commitId")));
				break;
			case "exec":
				handleExec(view, asString(message.get("id")), asStringList(message.get("command")));
				break;
			default:
				break;
		}
	}

	public void refreshAllLanguages() {
		String language = GitChordPreferences.effectiveLanguage();
		for (GitChordGui view : snapshotViews()) {
			view.postMessage(mapOf("type", "language", "language", language));
		}
	}

	private void openPage(PanelGroup group, String path, PageIntent intent) {
		runOnUiThread(() -> {
			IWorkbenchWindow window = activeWorkbenchWindow();
			if (window == null) {
				showError("No active Eclipse workbench window found.");
				return;
			}

			IWorkbenchPage page = window.getActivePage();
			if (page == null) {
				showError("No active Eclipse workbench page found.");
				return;
			}

			String secondaryId = secondaryIdForGroup(group);
			PanelState nextState = new PanelState(group, normalizePath(path), intent);
			synchronized (this) {
				statesBySecondaryId.put(normalizeSecondaryId(secondaryId), nextState);
			}

			try {
				IViewPart part = group.isRepo()
					? page.showView(GitChordGui.ID, secondaryId, IWorkbenchPage.VIEW_ACTIVATE)
					: page.showView(GitChordGui.ID);
				if (part instanceof GitChordGui) {
					((GitChordGui) part).applyState(nextState);
				}
			} catch (PartInitException e) {
				showError(e.getMessage());
			}
		});
	}

	private void handleOpenPageRequest(GitChordGui sourceView, PageOpenRequest request) {
		String path = normalizePath(request == null ? null : request.path());
		String type = request == null ? null : request.type();
		if ("global".equals(type)) {
			openPage(PanelGroup.global(), path, request.intent());
		} else if ("repo".equals(type)) {
			String repoRoot = request.repoRoot();
			if (repoRoot == null || repoRoot.isBlank()) {
				showError("No repository path was provided.");
			} else {
				openPage(PanelGroup.repo(repoRoot), path, request.intent());
			}
		} else if ("currentRepo".equals(type)) {
			openCurrentRepoPage(path, request.intent());
		} else {
			sourceView.navigate(path, request == null ? null : request.intent());
		}
	}

	private void handleSetLanguage(String language) {
		GitChordPreferences.setLanguage(language);
		refreshAllLanguages();
	}

	private void handleExec(GitChordGui view, String id, List<String> command) {
		if (id == null || command == null) {
			return;
		}

		if (!isGitChordCommand(command)) {
			view.postMessage(execResultMessage(id, new CommandResultPayload(
				1,
				"",
				"Rejected command: only git chord commands can be executed."
			)));
			return;
		}

		runInBackground(() -> {
			CommandResultPayload result = runGitChordCommand(command, workingDirectory(view.group()));
			view.postMessage(execResultMessage(id, result));
		});
	}

	private void refreshViewRepoContext(GitChordGui view) {
		if (view.group().isRepo()) {
			view.postMessage(mapOf("type", "repoContext", "repoRoot", view.group().repoRoot()));
			return;
		}

		runInBackground(() -> {
			String repoRoot = resolveCurrentRepoRoot();
			view.postMessage(mapOf("type", "repoContext", "repoRoot", repoRoot));
		});
	}

	private String initialRepoRoot(PanelGroup group) {
		return group.isRepo() ? group.repoRoot() : null;
	}

	private void handleOpenCommit(PanelGroup group, String commitId) {
		if (commitId == null || !COMMIT_ID_PATTERN.matcher(commitId).matches()) {
			showError("Invalid commit id.");
			return;
		}

		runInBackground(() -> {
			String repoRoot = group.isRepo() ? group.repoRoot() : resolveCurrentRepoRoot();
			if (repoRoot == null) {
				showError("No Git repository context found.");
				return;
			}

			CommandResultPayload result = runProcess(
				List.of(
					"git",
					"-C",
					repoRoot,
					"show",
					"--stat",
					"--patch",
					"--find-renames",
					"--find-copies",
					"--color=never",
					commitId
				),
				null,
				COMMIT_TIMEOUT_MILLIS
			);
			if (result.status() == 0) {
				openCommitDocument(commitId, result.stdout());
			} else {
				showError(result.stderr().isBlank() ? "Could not open commit." : result.stderr());
			}
		});
	}

	private void openCommitDocument(String commitId, String content) {
		runOnUiThread(() -> {
			try {
				Path tempFile = Files.createTempFile("git-chord-" + commitId.substring(0, Math.min(12, commitId.length())) + "-", ".diff");
				tempFile.toFile().deleteOnExit();
				Files.writeString(tempFile, content, StandardCharsets.UTF_8);
				IWorkbenchWindow window = activeWorkbenchWindow();
				if (window == null || window.getActivePage() == null) {
					showError("No active Eclipse workbench page found.");
					return;
				}
				IDE.openEditorOnFileStore(
					window.getActivePage(),
					EFS.getLocalFileSystem().fromLocalFile(tempFile.toFile())
				);
			} catch (Exception e) {
				showError(e.getMessage() == null ? "Could not create commit preview." : e.getMessage());
			}
		});
	}

	private CommandResultPayload runGitChordCommand(List<String> command, String cwd) {
		String configuredCommandPath = GitChordPreferences.effectiveCommandPath();
		List<String> resolvedCommand = new ArrayList<>();
		if (!configuredCommandPath.isBlank()) {
			resolvedCommand.add(configuredCommandPath);
			resolvedCommand.addAll(command.subList(2, command.size()));
		} else {
			resolvedCommand.addAll(command);
		}
		return runProcess(resolvedCommand, cwd, COMMAND_TIMEOUT_MILLIS);
	}

	private CommandResultPayload runProcess(List<String> command, String cwd, long timeoutMillis) {
		try {
			ProcessBuilder processBuilder = new ProcessBuilder(command);
			if (cwd != null && !cwd.isBlank()) {
				processBuilder.directory(new File(cwd));
			}

			Process process = processBuilder.start();
			Future<String> stdoutFuture = executorService.submit(streamReader(process.getInputStream()));
			Future<String> stderrFuture = executorService.submit(streamReader(process.getErrorStream()));
			boolean finished = process.waitFor(timeoutMillis, TimeUnit.MILLISECONDS);
			if (!finished) {
				process.destroyForcibly();
				return new CommandResultPayload(1, safeGet(stdoutFuture), "Command timed out.");
			}
			return new CommandResultPayload(process.exitValue(), safeGet(stdoutFuture), safeGet(stderrFuture));
		} catch (Exception e) {
			return new CommandResultPayload(1, "", e.getMessage() == null ? e.getClass().getSimpleName() : e.getMessage());
		}
	}

	private Callable<String> streamReader(InputStream inputStream) {
		return () -> new String(inputStream.readAllBytes(), StandardCharsets.UTF_8);
	}

	private String safeGet(Future<String> future) {
		try {
			return future.get(1, TimeUnit.SECONDS);
		} catch (Exception e) {
			return "";
		}
	}

	private boolean isGitChordCommand(List<String> command) {
		return command.size() >= 2 && "git".equals(command.get(0)) && "chord".equals(command.get(1));
	}

	private String resolveCurrentRepoRoot() {
		String editorPath = candidatePathForActiveEditor();
		if (editorPath != null) {
			String repoRoot = resolveRepoRootForPath(editorPath);
			if (repoRoot != null) {
				return repoRoot;
			}
		}

		String contextDirectory = GitChordPreferences.effectiveContextDirectory();
		if (!contextDirectory.isBlank()) {
			String repoRoot = resolveRepoRootForPath(contextDirectory);
			if (repoRoot != null) {
				return repoRoot;
			}
		}

		String workspacePath = workspacePath();
		return workspacePath == null ? null : resolveRepoRootForPath(workspacePath);
	}

	private String resolveRepoRootForPath(String path) {
		String candidatePath = normalizeCandidatePath(path);
		if (candidatePath == null) {
			return null;
		}

		CommandResultPayload result = runProcess(
			List.of("git", "-C", candidatePath, "rev-parse", "--show-toplevel"),
			null,
			REPO_RESOLVE_TIMEOUT_MILLIS
		);
		return result.status() == 0 ? result.stdout().trim() : null;
	}

	private String candidatePathForSelection(ISelection selection) {
		if (!(selection instanceof IStructuredSelection)) {
			return null;
		}

		Object firstElement = ((IStructuredSelection) selection).getFirstElement();
		return candidatePathForObject(firstElement);
	}

	private String candidatePathForObject(Object object) {
		if (object instanceof IResource) {
			return pathForResource((IResource) object);
		}

		if (object instanceof IAdaptable) {
			IResource resource = ((IAdaptable) object).getAdapter(IResource.class);
			if (resource != null) {
				return pathForResource(resource);
			}
		}

		if (object instanceof File) {
			return ((File) object).getAbsolutePath();
		}

		return null;
	}

	private String pathForResource(IResource resource) {
		IPath location = resource.getLocation();
		return location == null ? null : location.toOSString();
	}

	private String candidatePathForActiveEditor() {
		IWorkbenchWindow window = activeWorkbenchWindow();
		IWorkbenchPage page = window == null ? null : window.getActivePage();
		IEditorPart editor = page == null ? null : page.getActiveEditor();
		IEditorInput input = editor == null ? null : editor.getEditorInput();
		if (input == null) {
			return null;
		}

		IResource resource = input.getAdapter(IResource.class);
		if (resource != null) {
			return pathForResource(resource);
		}

		if (input instanceof IPathEditorInput) {
			IPath path = ((IPathEditorInput) input).getPath();
			return path == null ? null : path.toOSString();
		}

		if (input instanceof IURIEditorInput) {
			URI uri = ((IURIEditorInput) input).getURI();
			if (uri != null && "file".equals(uri.getScheme())) {
				return Path.of(uri).toString();
			}
		}

		return null;
	}

	private String normalizeCandidatePath(String path) {
		if (path == null || path.isBlank()) {
			return null;
		}

		Path candidate = Path.of(path);
		if (Files.isDirectory(candidate)) {
			return candidate.toString();
		}
		Path parent = candidate.getParent();
		return parent == null ? null : parent.toString();
	}

	private String workspacePath() {
		try {
			IPath workspaceLocation = ResourcesPlugin.getWorkspace().getRoot().getLocation();
			return workspaceLocation == null ? null : workspaceLocation.toOSString();
		} catch (Exception e) {
			return null;
		}
	}

	private String workingDirectory(PanelGroup group) {
		if (group.isRepo()) {
			return group.repoRoot();
		}

		String contextDirectory = GitChordPreferences.effectiveContextDirectory();
		return contextDirectory.isBlank() ? workspacePath() : contextDirectory;
	}

	private synchronized List<GitChordGui> snapshotViews() {
		return new ArrayList<>(viewsByGroupKey.values());
	}

	private IWorkbenchWindow activeWorkbenchWindow() {
		if (!PlatformUI.isWorkbenchRunning()) {
			return null;
		}

		IWorkbenchWindow window = PlatformUI.getWorkbench().getActiveWorkbenchWindow();
		if (window != null) {
			return window;
		}

		IWorkbenchWindow[] windows = PlatformUI.getWorkbench().getWorkbenchWindows();
		return windows.length == 0 ? null : windows[0];
	}

	private void runInBackground(Runnable runnable) {
		executorService.submit(runnable);
	}

	private void runOnUiThread(Runnable runnable) {
		Display display = Display.getDefault();
		if (display == null || display.isDisposed()) {
			return;
		}

		if (display.getThread() == Thread.currentThread()) {
			runnable.run();
		} else {
			display.asyncExec(runnable);
		}
	}

	private void showError(String message) {
		runOnUiThread(() -> {
			IWorkbenchWindow window = activeWorkbenchWindow();
			MessageDialog.openError(window == null ? null : window.getShell(), "Git Chord", message);
		});
	}

	private Map<String, Object> execResultMessage(String id, CommandResultPayload result) {
		return mapOf("type", "execResult", "id", id, "result", result.toPayload());
	}

	private Map<String, Object> mapOf(Object... items) {
		Map<String, Object> result = new LinkedHashMap<>();
		for (int i = 0; i + 1 < items.length; i += 2) {
			result.put(String.valueOf(items[i]), items[i + 1]);
		}
		return result;
	}

	private String normalizePath(String path) {
		return path == null || path.isBlank() ? "/" : path;
	}

	private String asString(Object value) {
		return value instanceof String ? (String) value : null;
	}

	private List<String> asStringList(Object value) {
		if (!(value instanceof List<?>)) {
			return null;
		}

		List<String> result = new ArrayList<>();
		for (Object item : (List<?>) value) {
			if (!(item instanceof String)) {
				return null;
			}
			result.add((String) item);
		}
		return result;
	}

	private String secondaryIdForGroup(PanelGroup group) {
		if (!group.isRepo()) {
			return null;
		}

		return "repo-" + sha256(group.repoRoot()).substring(0, 16);
	}

	private String normalizeSecondaryId(String secondaryId) {
		return secondaryId == null ? GLOBAL_SECONDARY_ID : secondaryId;
	}

	private String sha256(String value) {
		try {
			MessageDigest digest = MessageDigest.getInstance("SHA-256");
			byte[] bytes = digest.digest(value.getBytes(StandardCharsets.UTF_8));
			StringBuilder result = new StringBuilder();
			for (byte b : bytes) {
				result.append(String.format("%02x", b));
			}
			return result.toString();
		} catch (NoSuchAlgorithmException e) {
			return Integer.toHexString(value.hashCode());
		}
	}

}

File: intellij/src/main/kotlin/hu/webarticum/gitchord/gui/intellij/services/GitChordService.kt
package hu.webarticum.gitchord.gui.intellij.services

import com.google.gson.Gson
import com.google.gson.JsonSyntaxException
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.components.Service
import com.intellij.openapi.components.service
import com.intellij.openapi.util.Disposer
import com.intellij.openapi.fileEditor.FileEditorManager
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.Messages
import com.intellij.openapi.vfs.LocalFileSystem
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.openapi.wm.ToolWindow
import com.intellij.openapi.wm.ToolWindowManager
import com.intellij.ui.content.Content
import com.intellij.ui.content.ContentFactory
import hu.webarticum.gitchord.gui.intellij.model.CommandResultPayload
import hu.webarticum.gitchord.gui.intellij.model.PageIntent
import hu.webarticum.gitchord.gui.intellij.model.PageOpenRequest
import hu.webarticum.gitchord.gui.intellij.model.PanelGroup
import hu.webarticum.gitchord.gui.intellij.model.WebviewMessage
import hu.webarticum.gitchord.gui.intellij.settings.GitChordSettings
import hu.webarticum.gitchord.gui.intellij.toolWindow.GitChordPanel
import java.io.File
import java.io.InputStream
import java.nio.file.Files
import java.nio.file.Path
import java.util.concurrent.Callable
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

@Service(Service.Level.PROJECT)
class GitChordService(private val project: Project) {

    private val gson = Gson()

    private val panels = linkedMapOf<String, PanelRecord>()

    private var toolWindow: ToolWindow? = null

    fun attachToolWindow(toolWindow: ToolWindow) {
        runOnUiThread {
            this.toolWindow = toolWindow
            if (toolWindow.contentManager.contentCount == 0) {
                openPageInToolWindow(toolWindow, PanelGroup.global(), "/about", null)
            }
        }
    }

    fun openGlobalPage(path: String) {
        openPage(PanelGroup.global(), path, null)
    }

    fun openCurrentRepoPage(path: String = "/", intent: PageIntent? = null) {
        runInBackground {
            val repoRoot = resolveCurrentRepoRoot()
            invokeLater {
                if (repoRoot == null) {
                    showError("No Git repository context found.")
                    openPage(PanelGroup.global(), if (path == "/") "/repo-state" else path, intent)
                } else {
                    openPage(PanelGroup.repo(repoRoot), path, intent)
                }
            }
        }
    }

    fun openRepoStateForFile(file: VirtualFile?) {
        val candidatePath = file?.let { candidatePathForVirtualFile(it) }
        if (candidatePath == null) {
            showError("No file or directory selected.")
            return
        }

        runInBackground {
            val repoRoot = resolveRepoRootForPath(candidatePath)
            invokeLater {
                if (repoRoot == null) {
                    showError("The selected item is not inside a Git repository.")
                } else {
                    openPage(PanelGroup.repo(repoRoot), "/")
                }
            }
        }
    }

    fun handlePanelMessage(panel: GitChordPanel, rawMessage: String) {
        val message = try {
            gson.fromJson(rawMessage, WebviewMessage::class.java)
        } catch (e: JsonSyntaxException) {
            return
        } ?: return

        when (message.type) {
            "requestInitialState" -> panel.sendInitialState(initialRepoRoot(panel.group), language())
            "requestRepoContext" -> refreshPanelRepoContext(panel)
            "openPage" -> handleOpenPageRequest(panel, message.request)
            "setLanguage" -> handleSetLanguage(message.language)
            "openCommit" -> handleOpenCommit(panel.group, message.commitId)
            "exec" -> handleExec(panel, message)
        }
    }

    fun refreshAllLanguages() {
        val language = language()
        panels.values.forEach { it.panel.postMessage(mapOf("type" to "language", "language" to language)) }
    }

    private fun openPage(group: PanelGroup, path: String, intent: PageIntent? = null) {
        runOnUiThread {
            val toolWindow = toolWindow ?: ToolWindowManager.getInstance(project).getToolWindow(TOOL_WINDOW_ID)
            if (toolWindow == null) {
                showError("Git Chord tool window is not available.")
                return@runOnUiThread
            }

            this.toolWindow = toolWindow
            toolWindow.activate({ openPageInToolWindow(toolWindow, group, path, intent) }, true)
        }
    }

    private fun openPageInToolWindow(toolWindow: ToolWindow, group: PanelGroup, path: String, intent: PageIntent?) {
        val key = group.key()
        val existingRecord = panels[key]
        if (existingRecord != null) {
            toolWindow.contentManager.setSelectedContent(existingRecord.content)
            existingRecord.panel.navigate(path, intent)
            refreshPanelRepoContext(existingRecord.panel)
            existingRecord.panel.postMessage(mapOf("type" to "language", "language" to language()))
            return
        }

        val panel = GitChordPanel(this, group, path, intent)
        val content = ContentFactory.getInstance().createContent(panel.component, group.displayName(), false)
        val record = PanelRecord(group, panel, content)
        panels[key] = record
        content.setDisposer {
            panels.remove(key)
            Disposer.dispose(panel)
        }
        toolWindow.contentManager.addContent(content)
        toolWindow.contentManager.setSelectedContent(content)
    }

    private fun handleOpenPageRequest(sourcePanel: GitChordPanel, request: PageOpenRequest?) {
        val path = request?.path?.takeIf { it.isNotBlank() } ?: "/"
        when (request?.type) {
            "global" -> openPage(PanelGroup.global(), path, request.intent)
            "repo" -> {
                val repoRoot = request.repoRoot
                if (repoRoot.isNullOrBlank()) {
                    showError("No repository path was provided.")
                } else {
                    openPage(PanelGroup.repo(repoRoot), path, request.intent)
                }
            }
            "currentRepo" -> openCurrentRepoPage(path, request.intent)
            else -> sourcePanel.navigate(path, request?.intent)
        }
    }

    private fun handleSetLanguage(language: String?) {
        settings().setLanguage(language ?: GitChordSettings.DEFAULT_LANGUAGE)
        refreshAllLanguages()
    }

    private fun handleExec(panel: GitChordPanel, message: WebviewMessage) {
        val id = message.id ?: return
        val command = message.command ?: return
        if (!isGitChordCommand(command)) {
            panel.postMessage(execResultMessage(id, CommandResultPayload(
                1,
                "",
                "Rejected command: only git chord commands can be executed.",
            )))
            return
        }

        runInBackground {
            val result = runGitChordCommand(command, workingDirectory(panel.group))
            panel.postMessage(execResultMessage(id, result))
        }
    }

    private fun refreshPanelRepoContext(panel: GitChordPanel) {
        if (panel.group.type == PanelGroup.TYPE_REPO) {
            panel.postMessage(mapOf("type" to "repoContext", "repoRoot" to panel.group.repoRoot))
            return
        }

        runInBackground {
            val repoRoot = resolveCurrentRepoRoot()
            panel.postMessage(mapOf("type" to "repoContext", "repoRoot" to repoRoot))
        }
    }

    private fun initialRepoRoot(group: PanelGroup): String? {
        return if (group.type == PanelGroup.TYPE_REPO) {
            group.repoRoot
        } else {
            null
        }
    }

    private fun handleOpenCommit(group: PanelGroup, commitId: String?) {
        if (commitId == null || !COMMIT_ID_PATTERN.matches(commitId)) {
            showError("Invalid commit id.")
            return
        }

        runInBackground {
            val repoRoot = if (group.type == PanelGroup.TYPE_REPO) group.repoRoot else resolveCurrentRepoRoot()
            if (repoRoot == null) {
                invokeLater {
                    showError("No Git repository context found.")
                }
                return@runInBackground
            }

            val result = runProcess(
                listOf(
                    "git",
                    "-C",
                    repoRoot,
                    "show",
                    "--stat",
                    "--patch",
                    "--find-renames",
                    "--find-copies",
                    "--color=never",
                    commitId,
                ),
                null,
                COMMIT_TIMEOUT_MILLIS,
            )
            invokeLater {
                if (result.status == 0) {
                    openCommitDocument(commitId, result.stdout)
                } else {
                    showError(result.stderr.ifBlank { "Could not open commit." })
                }
            }
        }
    }

    private fun openCommitDocument(commitId: String, content: String) {
        val tempFile = Files.createTempFile("git-chord-${commitId.take(12)}-", ".diff")
        tempFile.toFile().deleteOnExit()
        Files.writeString(tempFile, content)

        val virtualFile = LocalFileSystem.getInstance().refreshAndFindFileByNioFile(tempFile)
        if (virtualFile == null) {
            showError("Could not create commit preview.")
            return
        }

        FileEditorManager.getInstance(project).openFile(virtualFile, true)
    }

    private fun runGitChordCommand(command: List<String>, cwd: String?): CommandResultPayload {
        val configuredCommandPath = settings().effectiveCommandPath()
        val resolvedCommand = if (configuredCommandPath.isNotBlank()) {
            listOf(configuredCommandPath) + command.drop(2)
        } else {
            command
        }
        return runProcess(resolvedCommand, cwd, COMMAND_TIMEOUT_MILLIS)
    }

    private fun runProcess(command: List<String>, cwd: String?, timeoutMillis: Long): CommandResultPayload {
        return try {
            val processBuilder = ProcessBuilder(command)
            if (!cwd.isNullOrBlank()) {
                processBuilder.directory(File(cwd))
            }

            val process = processBuilder.start()
            val executor = Executors.newFixedThreadPool(2)
            try {
                val stdoutFuture = executor.submit(Callable { readStream(process.inputStream) })
                val stderrFuture = executor.submit(Callable { readStream(process.errorStream) })
                val finished = process.waitFor(timeoutMillis, TimeUnit.MILLISECONDS)
                if (!finished) {
                    process.destroyForcibly()
                    return CommandResultPayload(1, stdoutFuture.safeGet(), "Command timed out.")
                }

                CommandResultPayload(process.exitValue(), stdoutFuture.safeGet(), stderrFuture.safeGet())
            } finally {
                executor.shutdownNow()
            }
        } catch (e: Exception) {
            CommandResultPayload(1, "", e.message ?: e::class.java.simpleName)
        }
    }

    private fun readStream(inputStream: InputStream): String {
        return inputStream.bufferedReader().use { it.readText() }
    }

    private fun <T> java.util.concurrent.Future<T>.safeGet(): T {
        return get(1, TimeUnit.SECONDS)
    }

    private fun isGitChordCommand(command: List<String>): Boolean {
        return command.size >= 2 && command[0] == "git" && command[1] == "chord"
    }

    private fun resolveCurrentRepoRoot(): String? {
        val editorFile = FileEditorManager.getInstance(project).selectedFiles.firstOrNull()
        val editorPath = editorFile?.let { candidatePathForVirtualFile(it) }
        if (editorPath != null) {
            resolveRepoRootForPath(editorPath)?.let { return it }
        }

        return project.basePath?.let { resolveRepoRootForPath(it) }
    }

    private fun resolveRepoRootForPath(path: String): String? {
        val candidatePath = normalizeCandidatePath(path) ?: return null
        val result = runProcess(
            listOf("git", "-C", candidatePath, "rev-parse", "--show-toplevel"),
            null,
            REPO_RESOLVE_TIMEOUT_MILLIS,
        )
        return if (result.status == 0) result.stdout.trim().takeIf { it.isNotBlank() } else null
    }

    private fun candidatePathForVirtualFile(file: VirtualFile): String {
        return if (file.isDirectory) file.path else file.parent?.path ?: file.path
    }

    private fun normalizeCandidatePath(path: String): String? {
        val candidate = Path.of(path)
        return when {
            Files.isDirectory(candidate) -> candidate.toString()
            candidate.parent != null -> candidate.parent.toString()
            else -> null
        }
    }

    private fun workingDirectory(group: PanelGroup): String? {
        return if (group.type == PanelGroup.TYPE_REPO) group.repoRoot else project.basePath
    }

    private fun execResultMessage(id: String, result: CommandResultPayload): Map<String, Any> {
        return mapOf("type" to "execResult", "id" to id, "result" to result)
    }

    private fun language(): String {
        return settings().effectiveLanguage()
    }

    private fun settings(): GitChordSettings {
        return ApplicationManager.getApplication().service()
    }

    private fun runInBackground(action: () -> Unit) {
        ApplicationManager.getApplication().executeOnPooledThread(action)
    }

    private fun invokeLater(action: () -> Unit) {
        ApplicationManager.getApplication().invokeLater(action)
    }

    private fun runOnUiThread(action: () -> Unit) {
        val application = ApplicationManager.getApplication()
        if (application.isDispatchThread) {
            action()
        } else {
            application.invokeLater(action)
        }
    }

    private fun showError(message: String) {
        runOnUiThread {
            Messages.showErrorDialog(project, message, "Git Chord")
        }
    }

    private data class PanelRecord(
        val group: PanelGroup,
        val panel: GitChordPanel,
        val content: Content,
    )

    companion object {
        const val TOOL_WINDOW_ID = "GitChordWindow"
        private const val COMMAND_TIMEOUT_MILLIS = 30_000L
        private const val COMMIT_TIMEOUT_MILLIS = 30_000L
        private const val REPO_RESOLVE_TIMEOUT_MILLIS = 5_000L
        private val COMMIT_ID_PATTERN = Regex("^[0-9a-fA-F]{7,40}$")
    }
}
